@yenda There is a problem [in the CI](https://ci.status.im/blue/organizations/jenkins/status-go%2Fprs%2Ftests/detail/PR-1742/3/pipeline).

```
--- FAIL: TestLoginWithKey (0.06s)
```

It seems legit.

--- a/api/backend_test.go
+++ b/api/backend_test.go
@@ -565,9 +565,11 @@ func TestLoginWithKey(t *testing.T) {
 	utils.Init()
 
 	b := NewGethStatusBackend()
-	pkey, err := crypto.GenerateKey()
+	chatKey, err := crypto.GenerateKey()
 	require.NoError(t, err)
-	keyUIDHex := sha256.Sum256(crypto.FromECDSAPub(&pkey.PublicKey))
+	walletKey, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	keyUIDHex := sha256.Sum256(crypto.FromECDSAPub(&chatKey.PublicKey))
 	keyUID := types.EncodeHex(keyUIDHex[:])
 	main := multiaccounts.Account{
 		KeyUID: keyUID,
@@ -577,14 +579,14 @@ func TestLoginWithKey(t *testing.T) {
 	defer os.Remove(tmpdir)
 	conf, err := params.NewNodeConfig(tmpdir, 1777)
 	require.NoError(t, err)
-	keyhex := hex.EncodeToString(crypto.FromECDSA(pkey))
+	keyhex := hex.EncodeToString(crypto.FromECDSA(chatKey))
 
 	require.NoError(t, b.AccountManager().InitKeystore(conf.KeyStoreDir))
 	b.UpdateRootDataDir(conf.DataDir)
 	require.NoError(t, b.OpenAccounts())
 
-	address := crypto.PubkeyToAddress(pkey.PublicKey)
-	require.NoError(t, b.SaveAccountAndStartNodeWithKey(main, "test-pass", conf, []accounts.Account{{Address: address}}, keyhex))
+	address := crypto.PubkeyToAddress(walletKey.PublicKey)
+	require.NoError(t, b.SaveAccountAndStartNodeWithKey(main, "test-pass", conf, []accounts.Account{{Address: address, Wallet: true}}, keyhex))
 	require.NoError(t, b.Logout())
 	require.NoError(t, b.StopNode())
 
@@ -595,5 +597,5 @@ func TestLoginWithKey(t *testing.T) {
 	}()
 	extkey, err := b.accountManager.SelectedChatAccount()
 	require.NoError(t, err)
-	require.Equal(t, crypto.PubkeyToAddress(pkey.PublicKey), extkey.Address)
+	require.Equal(t, crypto.PubkeyToAddress(chatKey.PublicKey), extkey.Address)
 }

--- a/api/geth_backend.go
+++ b/api/geth_backend.go
@@ -208,6 +208,8 @@ func (b *GethStatusBackend) SaveAccountAndStartNodeWithKey(acc multiaccounts.Acc
 
 // StartNodeWithKey instead of loading addresses from database this method derives address from key
 // and uses it in application.
+// TODO: we should use a proper struct with optional values instead of duplicating the regular functions
+// with small variants for keycard, this created too many bugs
 func (b *GethStatusBackend) startNodeWithKey(acc multiaccounts.Account, password string, keyHex string) error {
 	err := b.ensureAppDBOpened(acc, password)
 	if err != nil {
@@ -220,7 +222,15 @@ func (b *GethStatusBackend) startNodeWithKey(acc multiaccounts.Account, password
 	if err := logutils.OverrideRootLogWithConfig(conf, false); err != nil {
 		return err
 	}
-
+	accountsDB := accounts.NewDB(b.appDB)
+	walletAddr, err := accountsDB.GetWalletAddress()
+	if err != nil {
+		return err
+	}
+	watchAddrs, err := accountsDB.GetAddresses()
+	if err != nil {
+		return err
+	}
 	chatKey, err := ethcrypto.HexToECDSA(keyHex)
 	if err != nil {
 		return err
@@ -230,11 +240,11 @@ func (b *GethStatusBackend) startNodeWithKey(acc multiaccounts.Account, password
 		return err
 	}
 	b.accountManager.SetChatAccount(chatKey)
-	chatAcc, err := b.accountManager.SelectedChatAccount()
+	_, err = b.accountManager.SelectedChatAccount()
 	if err != nil {
 		return err
 	}
-	b.accountManager.SetAccountAddresses(chatAcc.Address)
+	b.accountManager.SetAccountAddresses(walletAddr, watchAddrs...)
 	err = b.injectAccountIntoServices()
 	if err != nil {
 		return err

